### PR TITLE
Added check for existence of Location HTTP header before using it

### LINKED ISSUE #6543: WebRequestPSCmdlet does not check existence of Location header when Authorization is present
The code introduced in https://github.com/PowerShell/PowerShell/pull/3885 seems to have introduced a regression bug which was quite hard to identify as the error message points to issues with the URI and not to anything related to redirection. [The code introduced in the pull request initializes an URI using the Location Header, without first checking if the Location Header is present in the response of a redirect.](https://github.com/PowerShell/PowerShell/blob/d20d53eaacafd4645ea668bd46e62093d3f3e503/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L1310). The [HTTP RFC 7231](https://tools.ietf.org/html/rfc7231#section-6.4) does not specify the Location Header to be mandatory and thus it's existence must be checked.

I discovered this while checking strange PowerShell errors when issuing a GET or HEAD request against a AWS S3 Bucket while connecting to the wrong region. AWS S3 does respond with HTTP 301 but without a Location header.

The issue only occurs if the Server responds with a redirect HTTP Status Code 3XX and without the Location header and if the client has specified the Authorization Header in the initial request.

The issue can probably be fixed by modifying [line 1290](https://github.com/PowerShell/PowerShell/blob/d20d53eaacafd4645ea668bd46e62093d3f3e503/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L1290) to `if (stripAuthorization && IsRedirectCode(response.StatusCode) && response.Headers.Location)`.

I'd be glad to create a pull request to fix this issue if my analyzis is correct and the proposed change is the correct way to fix the issue.

Actual behavior
---------------

The following simple HEAD request demonstrates the unexpected error message:

```powershell
Invoke-WebRequest -Uri https://s3.amazonaws.com/test-301-redirection/ -Method HEAD -Headers @{Authorization="test"}                       
Invoke-WebRequest : Value cannot be null.
Parameter name: uri
At line:1 char:1
+ Invoke-WebRequest -Uri https://s3.amazonaws.com/test-301-redirection/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [Invoke-WebRequest], ArgumentNullException
+ FullyQualifiedErrorId : System.ArgumentNullException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

Expected behavior
-----------------

The response should be

```powershell
Invoke-WebRequest -Uri https://s3.amazonaws.com/test-301-redirection/ -Method HEAD -Headers @{Authorization="test"}                                                          
Invoke-WebRequest : Response status code does not indicate success: 301 (Moved Permanently).
At line:1 char:1
+ Invoke-WebRequest -Uri https://s3.amazonaws.com/test-301-redirection/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (Method: HEAD, R...rShell/6.0.2
}:HttpRequestMessage) [Invoke-WebRequest], HttpResponseException
+ FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

A workaround is, to specify the `PreserveAuthorizationOnRedirect` Parameter which will skip the section introduced in https://github.com/PowerShell/PowerShell/pull/3885 :

```powershell
Invoke-WebRequest -Uri https://s3.amazonaws.com/test-301-redirection/ -Method HEAD -Headers @{Authorization="test"} -PreserveAuthorizationOnRedirect                                                                                                                                                                                             
Invoke-WebRequest : Response status code does not indicate success: 301 (Moved Permanently).
At line:1 char:1
+ Invoke-WebRequest -Uri https://s3.amazonaws.com/test-301-redirection/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (Method: HEAD, R...rShell/6.0.2
}:HttpRequestMessage) [Invoke-WebRequest], HttpResponseException
+ FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

Environment data
----------------

```powershell
$PSVersionTable                                                                                                                           

Name                           Value                                                                                                                                                                  
----                           -----                                                                                                                                                                  
PSVersion                      6.0.2                                                                                                                                                                  
PSEdition                      Core                                                                                                                                                                   
GitCommitId                    v6.0.2                                                                                                                                                                 
OS                             Darwin 17.4.0 Darwin Kernel Version 17.4.0: Sun Dec 17 09:19:54 PST 2017; root:xnu-4570.41.2~1/RELEASE_X86_64                                                          
Platform                       Unix                                                                                                                                                                   
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                
PSRemotingProtocolVersion      2.3                                                                                                                                                                    
SerializationVersion           1.1.0.1                                                                                                                                                                
WSManStackVersion              3.0                                           
```

### PR DESCRIPTION
**Pull request by ffeldhaus**
## PR Summary

Fix #6543 

The [HTTP RFC does not require a Location header to be present for redirects](https://tools.ietf.org/html/rfc7231#section-6.4), thus it is required to check if the Location header is returned before using it. This pull request adds a check for the Location header.

A test has been implemented to ensure that the issue does not show up again.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
- **Testing - New and feature**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by markekraus** [comment]
Hi @ffeldhaus ! Can you please add a test for this for bot `Invoke-WebRequest` and `Invoke-RestMethod` so we can ensure future changes don't cause a regression?

Also, for this PR, make sure you include `[feature]` in the commit messages to ensure the web cmdlet tests run.

Thanks!

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs [hunk 1]
```diff
@@ -1287,7 +1287,7 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
             _cancelToken = new CancellationTokenSource();
             HttpResponseMessage response = client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, _cancelToken.Token).GetAwaiter().GetResult();
 
-            if (stripAuthorization && IsRedirectCode(response.StatusCode))
+            if (stripAuthorization && IsRedirectCode(response.StatusCode) && response.Headers.Location)
```
Comment (markekraus) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs]: should be `response.Headers.Location != null`

**Review by markekraus — CHANGES_REQUESTED** [review]

**Comment by ffeldhaus** [comment]
Hi @markekraus thanks for the suggestion. I already fixed the initial commit, I should have tested it before submitting it. I now confirmed that the fix is working and committed the correct fix.

I would like to add a test, but as far as I understand WebListener, it always returns the Location header for Redirects. What would be the best way to make WebListener not return the Location header? If you can help me with the WebListener, I'd be glad to commit a test for both Cmdlets.

**Comment by markekraus** [comment]
@ffeldhaus Have a look at the `/Response/` endpoint https://github.com/PowerShell/PowerShell/tree/master/test/tools/WebListener#response 
You can have it return a redirect code without a Location header.

something like this

```powershell
$uri = Get-WebListenerUrl -Test Response -Query @{statuscode = 301}
{ Invoke-WebRequest -uri $uri -Headers @{Authorization = 'foo'} -ErroractionStop } |
    Should -Throw -ErrorId 'WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand'
```

**Comment by markekraus** [comment]
@ffeldhaus It looks like an unintentional change was made in your last commit that deleted an unrelated file. Can you please revert the file?

**Comment by ffeldhaus** [comment]
I changed the status code to int and are using -Match now isntead of -Be for checking the Exception message. The reason is, that the Exception Message contains the Status Code with a space, e.g. `Temporary Redirect` and not `TemporaryRedirect` which is provided in `$redirectTests`.

The tests are still failing because `$redirectTests` contains an entry `relative` which is not a valid Status Code. I'm not sure how to handle this. In my opinion it would be better to not include `relative` in the list of `$redirectTests` or at least separate it into Status Codes to test for redirect and an additional `relative` redirect test.

**Comment by markekraus** [comment]
@ffeldhaus for the `relative` issue, you could do 

```powershell
# Skip relative test as it is not a valid response type.
if ($redirectType -eq 'relative'){ return }
```

For the error message. I'm not certain we should even test it. Ideally, that message would be localized so it would fail in another language. I think the other tests are sufficient to show we are getting the error condition we desire and not the unrelated one.

**Comment by ffeldhaus** [comment]
@markekraus Thanks for all the help! I'm a bit annoyed at myself for doing quite a few mistakes in such an easy fix and that you basically had to tell me what to do. For the next time, I hope I'm now better equipped to make a change.

I think the pull request should be good to merge now.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 2]
```diff
@@ -822,6 +822,24 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
...
     }
 
+    It "Validates Invoke-WebRequest handles responses without Location header correctly for redirect: <redirectType>" -TestCases $redirectTests {
+        param($redirectType, $redirectedMethod)
+        # Skip relative test as it is not a valid response type.
+        if ($redirectType -eq 'relative'){ return }
+
+        # PowerShell should throw an HTTP Response Exception
+        # for a redirect response which does not contain a Location header
+        # the correct redirect status code should be included in the exception message
```
Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: We should include a note about  this requiring a Authorization header. Something like

```powershell
# When an Authorization request header is present,
# and -PreserveAuthorizationOnRedirect is not present,
# PowerShell should throw an HTTP Response Exception
# for a redirect response which does not contain a Location response header.
# The correct redirect status code should be included in the exception.
```

It's a mouth full, but at least people will understand why we are doing what seems like a superfluous set of tests.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 3]
```diff
@@ -2137,6 +2155,24 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
...
     }
 
+    It "Validates Invoke-RestMethod handles responses without Location header correctly for redirect: <redirectType>" -TestCases $redirectTests {
+        param($redirectType, $redirectedMethod)
+        # Skip relative test as it is not a valid response type.
+        if ($redirectType -eq 'relative'){ return }
+
+        # PowerShell should throw an HTTP Response Exception
+        # for a redirect response which does not contain a Location header
+        # the correct redirect status code should be included in the exception message
```
Comment (markekraus) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Same as above.

**Review by markekraus — CHANGES_REQUESTED** [review]
One more change and I think we're good.

**Review by markekraus — APPROVED** [review]
LGTM

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 4]
```diff
@@ -822,6 +822,27 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
         $response.Content.Method | Should -Be $redirectedMethod
     }
 
+    It "Validates Invoke-WebRequest handles responses without Location header correctly for requests with Authorization header and redirect: <redirectType>" -TestCases $redirectTests {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: It is good for comment but very long for header. Please make shorter header and add the comment.
Below too.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 5]
```diff
@@ -822,6 +822,27 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
...
+        param($redirectType, $redirectedMethod)
+        # Skip relative test as it is not a valid response type.
+        if ($redirectType -eq 'relative'){ return }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Please add space after `)` - `if ($redirectType -eq 'relative') { return }`

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 [hunk 6]
```diff
@@ -2137,6 +2158,27 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
         $response.Content.Method | Should -Be $redirectedMethod
     }
 
+    It "Validates Invoke-RestMethod handles responses without Location header correctly for requests with Authorization header and redirect: <redirectType>" -TestCases $redirectTests {
+        param($redirectType, $redirectedMethod)
+        # Skip relative test as it is not a valid response type.
+        if ($redirectType -eq 'relative'){ return }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: Please add space after `)`.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 (hunk 4 above)
Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: @TravisEz13 I get conflicting feedback on that. For now, I think it's best if this follows the code around it. When I refactor the tests, I will address this globally.

Comment (ffeldhaus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I added the whitespace and shortened the test description by one word. It's now shorter than the surrounding tests. I agree that `Validates` is redundant, but as all the other tests include it, I include it for consistency. 

I don't agree that a test header should be short. It really must describe the test in a good way and removing anything (apart from `Validates`) would change the meaning of the test.

@markekraus already stated that he wants to cleanup the tests and he can consistently remove `Validates` for all tests.

Comment (TravisEz13) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1]: I'm ok with delaying the changes to a later refactoring.

**Comment by TravisEz13** [comment]
LGTM

**Review by markekraus — APPROVED** [review]

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@ffeldhaus Thanks for your contribution! Welcome back with new ones!

**Pull request merged**